Tests/DatadogTests/Datadog/Mocks/ServerMock.swift

```diff
@@ -0,0 +1,167 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import Foundation
+import XCTest
+
+private class ServerMockProtocol: URLProtocol {
+    override class func canInit(with request: URLRequest) -> Bool {
+        return true
+    }
+
+    override class func canonicalRequest(for request: URLRequest) -> URLRequest {
+        // By the time this `canonicalRequest(for:)` is called, the original `URLRequest` is
+        // already transformed in `URLSession` by encoding `httpBody` into a stream and
+        // setting `Content-Length` HTTP header. This means that the `request` is not the one
+        // that we originally sent. Here, we transform it back from `httpBodyStream` to `httpBody`.
+        if let httpBodyStream = request.httpBodyStream {
+            let contentLength = Int(request.allHTTPHeaderFields!["Content-Length"]!)!
+
+            var canonicalRequest = URLRequest(url: request.url!)
+            canonicalRequest.httpMethod = request.httpMethod
+            canonicalRequest.allHTTPHeaderFields = request.allHTTPHeaderFields
+            canonicalRequest.httpBody = httpBodyStream.readAllBytes(expectedSize: contentLength)
+            return canonicalRequest
+        } else {
+            return request
+        }
+    }
+
+    override func startLoading() {
+        guard let serverMock = ServerMock.activeInstance else {
+            preconditionFailure("Request was started while no `ServerMock` instance is active.")
+        }
+
+        if let response = serverMock.mockedResponse {
+            client?.urlProtocol(self, didReceive: response, cacheStoragePolicy: .notAllowed)
+        }
+        if let data = serverMock.mockedData {
+            client?.urlProtocol(self, didLoad: data)
+        }
+        if let error = serverMock.mockedError {
+            client?.urlProtocol(self, didFailWithError: error)
+        }
+
+        client?.urlProtocolDidFinishLoading(self)
+
+        DispatchQueue.main.async {
+            serverMock.record(newRequest: self.request)
+        }
+    }
+
+    override func stopLoading() {
+        precondition(ServerMock.activeInstance != nil, "Request was stopped while no `ServerMock` instance is active.")
+    }
+}
+
+/// All unit tests use this shared `URLSession`.
+private let sharedURLSession: URLSession = {
+    let configuration = URLSessionConfiguration.ephemeral
+    configuration.protocolClasses = [ServerMockProtocol.self]
+    return URLSession(configuration: configuration)
+}()
```

Comment:
very nice 👍 

maybe we could rename it to smth like `sharedTestsURLSession` (or anything with `test` in it) to help with its discoverability while writing new test cases?

Reply:
I can rename it for clarity in this file 👍. Although this won't impact the discoverability, as this is defined `private`:
```swift
private let sharedURLSession: URLSession = ...
```
and this instance is later exposed internally by `ServerMock`:
```swift
class ServerMock {
   let urlSession: URLSession = sharedURLSession
}
```

Reply:
sorry i missed that it was `private`, yes you are right 👍